### ISSUE TITLE
[unstable option] trailing_comma

### ISSUE THREAD

**Issue by scampi** [rust-lang/rustfmt#3379]
Tracking issue for unstable option: trailing_comma

**Comment by c02y** [comment]
The three code snippets ("Always", "Never", "Vertical") of https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#trailing_comma are exactly the same.

**Comment by calebcartwright** [comment]
Thank you for sharing @c02y. However, that is not correct, as those three snippets do indeed have variance with trailing commas as designed (check for the absence/presence of the trailing comma after the `sit` in the first line and after `adipiscing`).

**Comment by c02y** [comment]
Yeah, sorry, I didn't notice that, it will be better to highlight or add an arrow on those differences inside the code snippet.

**Comment by nab-os** [comment]
Hi, is there any reason why this feature is still considered unstable ?
I can't find any opened issue mentioning it.

**Comment by dcow** [comment]
I just tried using this feature and it adds trailing commas to function argument lists at the call site. This seems way too broad. Is this intended?

**Comment by calebcartwright** [comment]
@dcow - You didn't specify, but I'm going to assume you are using the `Always` variant of the [trailing_comma](https://rust-lang.github.io/rustfmt/?version=v1.4.38&search=#trailing_comma) option. In which case, the answer is yes, that is intended. Feel free to suggest changes to the associated documentation for that variant if you think it can be improved or found it unclear, but Always indeed means always and I can't imagine changing that.

You're also free to open a request for an additional variant if you'd like to try to codify the heuristics, though I'll go ahead and note upfront that's not something we'd be able to allocate any of our bandwidth to working on, and would be reliant upon the community to implement.

However, I'm not really sure I follow the logic of how a function signature differs from a call site of said function in this context. One of the most common, but not exclusive, reasons people like utilizing trailing commas is to avoid extra diff noise in version control when something changes particularly in multiline cases, regardless of whether that's an additional parameter in a function (which impacts both the sig and call sites), new members in tuples or structs, variants, etc.. That's why the default value is `Vertical`, but `Always` exists for those that really want the commas everywhere.

**Comment by LollipopFt** [comment]
hi, may I know when this will be stabilised? just asking.

**Comment by calebcartwright** [comment]
> hi, may I know when this will be stabilised? just asking.

https://github.com/rust-lang/rustfmt/issues/3377#issuecomment-1135281548

**Comment by externl** [comment]
I'm getting the same behavior as @dcow with the default "Vertical" setting when calling macros. For example I see this

```rust
let message = format!(
     "file is using {} by default", 
    "LF",
);
```

is formatted to:

```rust
let message = format!("file is using {} by default", "LF",);
```


Something like:

```rust
let message = my_format(
    "file is using ", 
    "LF by default",
);
```

formats correctly to:

```rust
let message = my_format("file is using ", "LF by default");
```

Not sure if this is a bug or another settings I'm missing?

**Comment by externl** [comment]
Just found this relevant comment: https://github.com/rust-lang/rustfmt/issues/4956#issuecomment-914822481.

**Comment by calebcartwright** [comment]
You beat me to it @externl :smile:  but yes macro calls are an entirely different story, and it's intentional that rustfmt isn't stripping user-added commas in those cases for reasons detailed in some of those (transitively) linked issues
